### PR TITLE
improve efficiency and security of WebRTC signalling

### PR DESCRIPTION
**Pull request by jamesdiacono**
Previously, signalling was symmetric. Connecting and listening parties both used the signalling server in the same way - by subscribing to signals destined for a name. The signalling server blindly trusted subscribing parties to give their real name, making it trivial to eavesdrop on signals intended for others. While this was not a security risk per se, it did make a DoS attack trivial to implement. (This approach was also very inefficient for listening parties, because often a bunch of identical connections would be made to the signalling server and they would all receive the same signals.)

While there is no easy way for the signalling server to verify a party's name (because a name is an RTCCertificate fingerprint), it is possible to demand credentials from parties attempting to listen while providing an unhindered endpoint for parties attempting to connect. (The credentials are a static password, "uFork", until we come up with something better.)

This commit introduces asymmetric signalling: there are now different endpoints for connecting and listening. I also managed to save a round trip by encoding the SDP offer in a URL, which is nice.

### CODE REVIEW

**Comment by alanhkarp** [comment]
What you've done is fine, but I'd like to point out an approach that goes further.  [User Managed Access](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html) (UMA, ) uses a server that enforces a subscriber's access policy.  In your case, a listener could specify a  policy, such as that the connecting party must prove membership in some organization.  The signaling server would only hand out an access token to the listener to those who can prove they meet the policy requirements.  That's probably more complexity than you should deal with now, but you might want to keep open the possibility of adding it later.

**Comment by jamesdiacono** [comment]
Would the listener trust their signalling server to verify the connecting party's organizational membership, or do it themself? Or both?

Is the benefit of having it done by the signalling server that resources of the listening party are not wasted establishing doomed connections?

**Comment by alanhkarp** [comment]
> Would the listener trust their signalling server to verify the connecting party's organizational membership, or do it themself? Or both?

With the UMA approach, the listener is expressing a trust relationship with the signaling server when the listener chooses to let the signaling server enforce the listener's access policy.

> Is the benefit of having it done by the signalling server that resources of the listening party are not wasted establishing doomed connections?

That and the fact that the listener can be offline when the access decision gets made.

**Comment by jamesdiacono** [comment]
Since we are forced to host a publicly addressable server to use WebRTC, it could make sense for it to do more than just signalling. However, that may cause the semantics of the WebRTC transport to diverge from the other transports (such as the TLS transport, for which I would prefer not to introduce additional servers).

Perhaps it would be preferable to verify membership at the actor level. The listening party's greeter actor is provided with the connecting party's verified secure name, allowing it to enforce its own security policy before handing out capabilities. What do you think?

**Comment by alanhkarp** [comment]
You already have a way to do 3-party introduction, so what you're talking about here is adding a configuration new to the set of configurations that are already connected with actor references.  As long as the vetting can be done without knowing the behavior of the listener,  I think it's best to centralize the vetting in the signaling server.  Otherwise, you're relying on every greeter actor to have the best security.  You can then have the greeter actor apply any configuration-specific rules as an additional filter.  

Not everyone does it this way.  One of the standards groups I lurk on assumes no trust in their equivalent of the signaling server.  It's a design decision.

It's probably best to design for something more general than just names even though you'll implement something simple at first.  A relatively new W3 standard is something called verifiable credentials.  A lot of thought has gone into them, so you should probably take a look.

**Comment by jamesdiacono** [comment]
I like the idea that vetting could be done without knowing the behavior of the listener, though I'm not sure how that would work for non-p2p transports that don't use a signalling server.

Bear in mind that not every AWP party needs to expose a greeter actor. A system of non-greeting parties could be shielded by a single greeting party that enforced a security policy, refusing to invite unauthorized parties into the system. Possible advantages of this approach is that a) the entirety of the application lives in actor-land, and b) an actor configuration could use multiple transports (WebRTC, TLS, etc) with no additional work.

I had a look at verifiable credentials, they seem like a good idea. An AWP "name" is a pretty abstract concept, so it's conceivable that an AWP transport might use VCs as names.

**Comment by alanhkarp** [comment]
These are all design decisions you'll have to make.  There's an argument for the equivalent of a signaling server even if the transport doesn't require one.  UMA used one to manage user policies; waterken used one to manage rotating keys. 

I like the approach in your second paragraph.

**Comment by jamesdiacono** [comment]
I'm very much hoping that features like key rotation can be built on top of what we have now, though I've not given it much thought.

Currently, a party's identity roughly corresponds to a keypair. A person might operate several parties each with unique identities (for example, a party on their laptop and a party on their phone). I'm expecting that a party's identity will be short-lived, but can be certified (owned?) by a longer-lived keypair, perhaps representing a person or organization.

This expectation is reflected in the use of RTCCertificates as identities for the WebRTC transport's parties. RTCCertificate objects wrap a keypair that can not be accessed directly. RTCCertificates can not be imported, exported, used to sign, or used to encrypt anything. They expire after one month. There are draconian limitations, but perhaps we can use them as strengths.

On the plus side, RTCCertificates can be persisted in the browser's IndexedDB. And they expose a "fingerprint" that we use as the party's secure name. This fingerprint could be signed by a person's long-lived private key, and that signature used to verify ownership. Ownership could be checked by the greeter responsible for enforcing the security policy.

In short, I'm expecting that there will be at least two "levels" of identity, and that may allow us to keep the transports simple.

**Comment by alanhkarp** [comment]
You should look at the [Decentralized Identifiers](https://www.w3.org/TR/did-core/) (DID) work.  It deals with several of the issues you mention, such as personas and key rotation.

**Comment by jamesdiacono** [comment]
Thanks!
